url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/org/jbox2d/dynamics/Island.java

```diff
@@ -0,0 +1,463 @@
+raint structures. The body velocities/positions are held in compact, temporary
+arrays to increase the number of cache hits. Linear and angular velocity are
+stored in a single array since multiple arrays lead to multiple misses.
+*/
+
+/*
+2D Rotation
+
+R = [cos(theta) -sin(theta)]
+    [sin(theta) cos(theta) ]
+
+thetaDot = omega
+
+Let q1 = cos(theta), q2 = sin(theta).
+R = [q1 -q2]
+    [q2  q1]
+
+q1Dot = -thetaDot * q2
+q2Dot = thetaDot * q1
+
+q1_new = q1_old - dt * w * q2
+q2_new = q2_old + dt * w * q1
+then normalize.
+
+This might be faster than computing sin+cos.
+However, we can compute sin+cos of the same angle fast.
+*/
+
+// pooled locally, not thread-safe
+// updated to rev 100
+/**
+ * This is an internal class.
+ * @author Daniel Murphy
+ */
+public class Island {
+	
+	public ContactListener m_listener;
+
+	public Body[] m_bodies;
+	public Contact[] m_contacts;
+	public Joint[] m_joints;
+
+	public Position[] m_positions;
+	public Velocity[] m_velocities;
+	
+	public int m_bodyCount;
+	public int m_jointCount;
+	public int m_contactCount;
+	
+	public int m_bodyCapacity;
+	public int m_contactCapacity;
+	public int m_jointCapacity;
+	
+	public int m_positionIterationCount;
+	
+	public Island(){
+		
+	}
+	
+	public void init(int bodyCapacity, int contactCapacity, int jointCapacity, ContactListener listener){
+		m_bodyCapacity = bodyCapacity;
+		m_contactCapacity = contactCapacity;
+		m_jointCapacity = jointCapacity;
+		m_bodyCount = 0;
+		m_contactCount = 0;
+		m_jointCount = 0;
+		
+		m_listener = listener;
+		
+		if(m_bodies == null || m_bodyCapacity > m_bodies.length){
+			m_bodies = new Body[m_bodyCapacity];
+		}
+		if(m_joints == null || m_jointCapacity > m_joints.length){
+			m_joints = new Joint[m_jointCapacity];			
+		}
+		if(m_contacts == null || m_contactCapacity > m_contacts.length){
+			m_contacts = new Contact[m_contactCapacity];
+		}
+		
+		// dynamic array
+		if(m_velocities == null || m_bodyCapacity > m_velocities.length){
+		  final Velocity[] old = m_velocities == null ? new Velocity[0] : m_velocities;
+			m_velocities = new Velocity[m_bodyCapacity];
+			System.arraycopy(old, 0, m_velocities, 0, old.length);
+			for(int i=old.length; i<m_velocities.length; i++){
+				m_velocities[i] = new Velocity();
+			}
+		}
+		
+		// dynamic array
+		if(m_positions == null || m_bodyCapacity > m_positions.length){
+		  final Position[] old = m_positions == null ? new Position[0] : m_positions;
+			m_positions = new Position[m_bodyCapacity];
+			System.arraycopy(old, 0, m_positions, 0, old.length);
+			for(int i=old.length; i<m_positions.length; i++){
+				m_positions[i] = new Position();
+			}
+		}
+	}
+	
+	public void clear(){
+		m_bodyCount = 0;
+		m_contactCount = 0;
+		m_jointCount = 0;
+	}
+	
+	private final Vec2 temp = new Vec2();
+	private final ContactSolver contactSolver = new ContactSolver();
+	private final Vec2 translation = new Vec2();
+	
+	public void solve(TimeStep step, Vec2 gravity, boolean allowSleep){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 32. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 238-276, lines 279-293, lines 296-369, lines 371-406 into separate methods.